chat-android/src/main/java/com/ably/chat/Messages.kt

```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
 )
 
 interface MessagesSubscription : Cancellation {
-    suspend fun getPreviousMessages(queryOptions: QueryOptions): PaginatedResult<Message>
+    suspend fun getPreviousMessages(start: Long? = null, end: Long? = null, limit: Int = 100): PaginatedResult<Message>
 }
 
-class DefaultMessages(
+internal class DefaultMessagesSubscription(
+    private val chatApi: ChatApi,
     private val roomId: String,
-    private val realtimeClient: RealtimeClient,
+    private val cancellation: Cancellation,
+    internal val fromSerialProvider: () -> DeferredValue<String>,
+) : MessagesSubscription {
+    override fun cancel() {
+        cancellation.cancel()
+    }
+
+    override suspend fun getPreviousMessages(start: Long?, end: Long?, limit: Int): PaginatedResult<Message> {
+        val fromSerial = fromSerialProvider().await()
+        val queryOptions = QueryOptions(start = start, end = end, limit = limit, orderBy = NewestFirst)
+        return chatApi.getMessages(
+            roomId = roomId,
+            options = queryOptions,
+            fromSerial = fromSerial,
+        )
+    }
+}
+
+internal class DefaultMessages(
+    private val roomId: String,
+    realtimeChannels: AblyRealtime.Channels,
     private val chatApi: ChatApi,
 ) : Messages {
 
+    private var listeners: Map<Messages.Listener, DeferredValue<String>> = emptyMap()
+
+    private var channelStateListener: ChannelStateListener
+
+    private var lock = Any()
+
     /**
      * the channel name for the chat messages channel.
      */
     private val messagesChannelName = "$roomId::\$chat::\$chatMessages"
 
-    override val channel: Channel
-        get() = realtimeClient.channels.get(messagesChannelName, ChatChannelOptions())
+    override val channel: Channel = realtimeChannels.get(messagesChannelName, ChatChannelOptions())
+
+    init {
+        channelStateListener = ChannelStateListener {
+            if (!it.resumed) updateChannelSerialsAfterDiscontinuity()
+        }
+        channel.on(ChannelState.attached, channelStateListener)
+    }
 
     override fun subscribe(listener: Messages.Listener): MessagesSubscription {
-        TODO("Not yet implemented")
+        val deferredChannelSerial = DeferredValue<String>()
+        addListener(listener, deferredChannelSerial)
+        val messageListener = PubSubMessageListener {
+            val pubSubMessage = it ?: throw AblyException.fromErrorInfo(
+                ErrorInfo("Got empty pubsub channel message", HttpStatusCodes.BadRequest, ErrorCodes.BadRequest),
+            )
+            val data = parsePubSubMessageData(pubSubMessage.data)
+            val chatMessage = Message(
+                roomId = roomId,
+                createdAt = pubSubMessage.timestamp,
```

Comment:
In JS we've tended towards native date types for `createdAt` - not sure what's better for Kotlin?

Reply:
I don't have strong opinion right now. Kotlin standard library doesn't have date type out of the box, and I still believe that one day we can provide  Kotlin multiplatform support for all our products :) I'll think about this, look how easy to use timestamps in example app. Also want to research how libraries with kotlin multiplatform support like apollo-graphql deals with this. For now I suggest keeping date as `Long`

Reply:
wdyt about https://github.com/Kotlin/kotlinx-datetime. It's a multiplatform datetime library

Reply:
If I’m not mistaken, it’s not stable yet, so I don’t think we should depend on it right now. We can decide this later—it’s not an architectural design decision that will cause issues for us later, and it will be easy to refactor whenever we want.